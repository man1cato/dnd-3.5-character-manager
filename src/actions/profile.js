import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
