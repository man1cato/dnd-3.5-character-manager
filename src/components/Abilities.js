import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
