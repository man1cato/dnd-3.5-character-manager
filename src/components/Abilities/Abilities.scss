@import '../../styles/base/settings';

.Abilities {  
   display: grid;
   grid-template-columns: repeat(4 1fr);
   grid-auto-flow: dense;
   grid-row-gap: $xs-size; 
   text-align: center;
   align-items: center;
}