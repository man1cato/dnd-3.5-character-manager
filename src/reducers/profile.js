export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
