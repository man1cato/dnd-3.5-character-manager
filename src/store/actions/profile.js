import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
