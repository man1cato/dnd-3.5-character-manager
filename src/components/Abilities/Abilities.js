import React, { Fragment, useState, useEffect } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import Counter from '../Counter/Counter'
import { calcAbilityMod, convertInputValue } from '../../utils/utils'
import './Abilities.scss'


const Abilities = ({ abilities, handleUpdate }) => {
	const scores = _.mapValues(abilities, (obj) => obj.score)
	const [tempScores, setTempScores] = useState(_.mapValues(abilities, (obj) => obj.tempScore))

	useEffect(() => {
		handleUpdate({ 
			abilities: _.mapValues(tempScores, (val, ability) => ({
				score: scores[ability],
				tempScore: val
			})) 
		})
	}, [tempScores])

	return (
		<div className="Abilities">
			<h5 className="grid__col1">Ability</h5>
			<h5 className="grid__col2">Score</h5>
			<h5 className="grid__col3">Temp Score</h5>
			<h5 className="grid__col4">Mod</h5>

			{_.map(['str', 'dex', 'con', 'int', 'wis', 'cha'], ability => {
				const score = scores[ability]
				const tempScore = tempScores[ability]
				return (
					<Fragment key={ability}>
						<div className="grid__col1">{ability.toUpperCase()}</div>
						<div className="grid__col2">{score}</div>
						<Counter
							className="grid__col3"
							value={convertInputValue(tempScore)}
							updateValue={value => setTempScores(update(tempScores, {
								[ability]: { $set: convertInputValue(value) }
							}))}
						/>
						<div 
							className="grid__col4"
							data-testid={ability + 'Mod'}
						>
							{!!tempScore ? calcAbilityMod(tempScore) : calcAbilityMod(score)}
						</div>
					</Fragment>
				)
			})}
		</div>
	)
}


export default Abilities