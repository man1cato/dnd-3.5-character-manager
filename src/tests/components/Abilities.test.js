import React from 'react'
import { shallow } from 'enzyme'

import Abilities from '../../components/Abilities'
import profile from '../fixtures/profile'


const profileData = {
  id: profile.id,
  ...profile.fields
}
const handleChange = jest.fn()

let wrapper
beforeEach(() => {
  wrapper = shallow(
    <Abilities
      handleChange={handleChange}
      abilities={profile.fields.abilities}
    />
  )
})

test('should render Abilities with profile data', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle input change', () => {
  const value = 14
  wrapper.find({ id: 'str' }).simulate('change', {
    target: { value }
  })
  expect(handleChange).toHaveBeenCalledWith({
    target: { value }
  })
})
