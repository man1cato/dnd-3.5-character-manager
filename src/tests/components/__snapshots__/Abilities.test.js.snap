// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Abilities with profile data 1`] = `
<div
  className="grid grid--abilities"
>
  <h5
    className="grid__col1"
  >
    Ability
  </h5>
  <div
    className="grid__col1"
    key="0"
  >
    STR
  </div>
  <div
    className="grid__col1"
    key="1"
  >
    DEX
  </div>
  <div
    className="grid__col1"
    key="2"
  >
    CON
  </div>
  <div
    className="grid__col1"
    key="3"
  >
    INT
  </div>
  <div
    className="grid__col1"
    key="4"
  >
    WIS
  </div>
  <div
    className="grid__col1"
    key="5"
  >
    CHA
  </div>
  <h5
    className="grid__col2"
  >
    Score
  </h5>
  <div
    className="grid__col2"
    key="0"
  >
    14
  </div>
  <div
    className="grid__col2"
    key="1"
  >
    18
  </div>
  <div
    className="grid__col2"
    key="2"
  >
    16
  </div>
  <div
    className="grid__col2"
    key="3"
  >
    20
  </div>
  <div
    className="grid__col2"
    key="4"
  >
    14
  </div>
  <div
    className="grid__col2"
    key="5"
  >
    11
  </div>
  <h5
    className="grid__col3"
  >
    Mod
  </h5>
  <div
    className="grid__col3"
    key="0"
  >
    2
  </div>
  <div
    className="grid__col3"
    key="1"
  >
    2
  </div>
  <div
    className="grid__col3"
    key="2"
  >
    3
  </div>
  <div
    className="grid__col3"
    key="3"
  >
    5
  </div>
  <div
    className="grid__col3"
    key="4"
  >
    2
  </div>
  <div
    className="grid__col3"
    key="5"
  >
    0
  </div>
  <h5
    className="grid__col4"
  >
    Temp Score
  </h5>
  <input
    className="grid__col4"
    id="str"
    key="0"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <input
    className="grid__col4"
    id="dex"
    key="1"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <input
    className="grid__col4"
    id="con"
    key="2"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <input
    className="grid__col4"
    id="int"
    key="3"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <input
    className="grid__col4"
    id="wis"
    key="4"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <input
    className="grid__col4"
    id="cha"
    key="5"
    name="abilities"
    onChange={[Function]}
    type="text"
  />
  <h5
    className="grid__col5"
  >
    Temp Mod
  </h5>
  <div
    className="grid__col5"
    key="0"
  />
  <div
    className="grid__col5"
    key="1"
  />
  <div
    className="grid__col5"
    key="2"
  />
  <div
    className="grid__col5"
    key="3"
  />
  <div
    className="grid__col5"
    key="4"
  />
  <div
    className="grid__col5"
    key="5"
  />
</div>
`;
